service/javadsl/kafka/server/src/main/scala/com/lightbend/lagom/internal/javadsl/broker/kafka/JavadslRegisterTopicProducers.scala

```diff
@@ -41,44 +50,59 @@ class JavadslRegisterTopicProducers @Inject() (
     tc      <- service.descriptor.topicCalls().asScala
-    topicCall = tc.asInstanceOf[TopicCall[AnyRef]]
+    topicCall = tc.asInstanceOf[TopicCall[BrokerMessage]]
   } {
     topicCall.topicHolder match {
       case holder: MethodTopicHolder =>
         val topicProducer = holder.create(service.service)
         val topicId       = topicCall.topicId
 
         topicFactory.create(topicCall) match {
-          case topicImpl: JavadslKafkaTopic[AnyRef] =>
+          case topicImpl: JavadslKafkaTopic[BrokerMessage] =>
             topicProducer match {
-              case tagged: TaggedOffsetTopicProducer[AnyRef, _] =>
+              case tagged: TaggedInternalTopic[BrokerMessage, Event] =>
                 val tags = tagged.tags.asScala.toIndexedSeq
 
-                val eventStreamFactory: (String, Offset) => Source[(AnyRef, Offset), _] = { (tag, offset) =>
-                  tags.find(_.tag == tag) match {
-                    case Some(aggregateTag) =>
-                      tagged
-                        .readSideStream(
-                          aggregateTag,
-                          OffsetAdapter.offsetToDslOffset(offset)
-                        )
-                        .asScala
-                        .map { pair =>
-                          pair.first -> OffsetAdapter.dslOffsetToOffset(pair.second)
+                val eventStreamFactory: EventStreamFactory[BrokerMessage] =
+                  tagged match {
+                    case producer: DelegatedTopicProducer[BrokerMessage, Event] =>
+                      val sourceFactory: (String, Offset) => Source[EventEnvelope, NotUsed] = (tag, offset) =>
+                        tags.find(_.tag == tag) match {
+                          case Some(aggregateTag) =>
+                            val lagomOffset = OffsetAdapter.offsetToDslOffset(offset)
+                            producer.persistentEntityRegistry
+                              .eventEnvelopeStream(aggregateTag, lagomOffset)
+                              .asScala
+                          case None => throw new RuntimeException("Unknown tag: " + tag)
                         }
-                    case None => throw new RuntimeException("Unknown tag: " + tag)
+                      DelegatedEventStreamFactory(sourceFactory, producer.userFlowAkka.asScala)
+                    case producer: TaggedOffsetTopicProducer[BrokerMessage, Event] =>
+                      ClassicLagomEventStreamFactory((tag, offset) =>
+                        tags.find(_.tag == tag) match {
+                          case Some(aggregateTag) =>
+                            val lagomOffset = OffsetAdapter.offsetToDslOffset(offset)
+                            producer
+                              .readSideStreamAkka(aggregateTag, lagomOffset)
+                              .asScala
+                          case None => throw new RuntimeException("Unknown tag: " + tag)
+                        }
+                      )
```

Comment:
This is an important change. Introducing `TopicProducer.fromTaggedEntity` required to build a new internal representation of a topic. There`s the classic `TaggedOffsetTopicProducer` (created by `TopicProducer.taggedStreamWithOffset`) and the new `DelegatedTopicProducer` (created by `TopicProducer.fromTaggedEntity`). 

In `Xxxxdsl.RegisterTopicProducers` there's now a pattern match doing custom preparations before invoking the projection driver (aka `Producer.startTaggedOffsetProducer`).

The pattern matching above uses the new ADT `EventStreamFactory`.